### PR TITLE
Add ticket opening from topology

### PR DESCRIPTION
**Pull request by djw8605**
Still need to be completed:

- [x] [Captcha](https://developers.google.com/recaptcha/docs/v3) from Google

This adds 2 new configuration values:

- `GOOGLE_CAPTCHA_KEY`: Registered CAPTCHA secret key.
- `FRESHDESK_API_KEY`: API key for FreshDesk. I believe each person has their own API key, and I don't know how to make it general for the entire project.

### CODE REVIEW

**Comment by djw8605** [comment]
Read for review.

File: src/app.py [hunk 1]
```diff
@@ -108,6 +110,65 @@ def contacts():
...
+                resource_tree = resource.get_tree(authorized = True)
+                contact_list = resource_tree['ContactLists']['ContactList']
+                for contact in contact_list:
+                    if contact["ContactType"] == "Administrative Contact":
+                        for inner_contact in contact['Contacts']['Contact']:
+                            cc_emails.add(inner_contact['Email'])
+
+    print(cc_emails)
+    ticket = {
+        'subject': "Test API ticket",
```
Comment (matyasselmeci) [inline src/app.py]: We should change the default subject before this goes live.

**Comment by bbockelm** [comment]
@djw8605 - can you attach some screenshots to the ticket so we can see what the UI looks like?

Also - I don't want to deploy any ticket UI that might ask for an X509 certificate.  Can we limit this to a specific vhost?

**Comment by brianhlin** [comment]
> @djw8605 - can you attach some screenshots to the ticket so we can see what the UI looks like?

Or better yet, deployment to the ITB?

**Comment by djw8605** [comment]
Multiple resources select with smart search.
image

It will add all of the "Admin" contacts for those resources.

**Comment by brianhlin** [comment]
Do we have a drop-down for VOs in case sites are having issues with a specific VO's?

**Comment by djw8605** [comment]
Nope, for now, this was specific to opening ticket against sites.

File: src/templates/ticket.html.j2 [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        <label for="requester">Requester Email</label>
+        <input type="email" name="requester" class="form-control" id="requester" required placeholder="Requester Email" pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$"></input>
+    </div>
+    <div class="form-group">
+        <label for="Subject">Subject</label>
+        <input type="text" name="subject" class="form-control" id="subject" required></input>
+    </div>
+    <div class="form-group">
+      <label for="ticket-message">
+        Ticket Content
```
Comment (brianhlin) [inline src/templates/ticket.html.j2]: I prefer "Description of the issue" or simply "Description"

File: src/templates/ticket.html.j2 [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    <div class="form-group">
+      <label for="resource-list">Resource</label>
+      <select name="resource-list" id="resource-list" multiple placeholder="Pick resources..."></select>
+    </div>
+    <div class="form-group">
+        <label for="requester">Requester Email</label>
+        <input type="email" name="requester" class="form-control" id="requester" required placeholder="Requester Email" pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$"></input>
+    </div>
+    <div class="form-group">
+        <label for="Subject">Subject</label>
```
Comment (brianhlin) [inline src/templates/ticket.html.j2]: "Summary of the issue" or "Summary"

File: src/templates/ticket.html.j2 [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+
+{% endblock %}
+
+{% block content %}
+<div class="container">
+
+  <form action="/submitTicket" method="POST">
+    <div class="form-group">
+      <label for="resource-list">Resource</label>
```
Comment (brianhlin) [inline src/templates/ticket.html.j2]: > Pick OSG Resource(s):

File: src/templates/ticket.html.j2 (hunk 2 above)
Comment (djw8605) [reply src/templates/ticket.html.j2]: I like "Description of the issue"

File: src/templates/ticket.html.j2 (hunk 4 above)
Comment (djw8605) [reply src/templates/ticket.html.j2]: That "Pick Resource(s)" is the placeholder text in the search box.  Should it be repeated?

Comment (brianhlin) [reply src/templates/ticket.html.j2]: Ah, I missed that. `Resource(s) is fine`

**Comment by djw8605** [comment]
Added a few sentences at the top:
image

**Comment by brianhlin** [comment]
That looks good. I forgot to ask, can we enable the Freshdesk "support" widget on this page, too?

**Comment by djw8605** [comment]
the freshdesk support widget is enabled.  You can see it in the screenshot on the right.

**Comment by brianhlin** [comment]
Oops, so it is. Could we get a link added to the main page under "Human-readable pages"?

File: src/app.py (hunk 1 above)
Comment (brianhlin) [reply src/app.py]: I believe this is fixed. Resolving

File: src/templates/base.html.j2 [hunk 5]
```diff
@@ -8,6 +8,9 @@
       href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" 
       integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE="
       crossorigin="anonymous" />
+<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"
```
Comment (matyasselmeci) [inline src/templates/base.html.j2]: What's the purpose of this change? If want_js is false, then javascript shouldn't get loaded...

Comment (djw8605) [reply src/templates/base.html.j2]: JQuery needs to be loaded before the script section so that `$` is defined.  I suppose this should be surrounded by the `want_js` if's.

Comment (matyasselmeci) [reply src/templates/base.html.j2]: Actually it looks like you added the Freshdesk Support Widget to base.html.j2 which is all JavaScript, so at this point I guess you might as well get rid of `want_js` altogether...

Comment (djw8605) [reply src/templates/base.html.j2]: Removed want_js

Comment (matyasselmeci) [reply src/templates/base.html.j2]: Can you edit contacts.html.j2 and generate_downtime_form.html.j2 to remove the `set want_js = true` lines?

**Comment by brianhlin** [comment]
Let's hold off on merging this until we clear up the issues mentioned here: https://support.opensciencegrid.org/a/tickets/8380

**Comment by djw8605** [comment]
The issues mentioned in that ticket are related to public ticket, not necessarily with this ticket creation.  I see them as unrelated.

File: src/app.py [hunk 6]
```diff
@@ -108,6 +112,69 @@ def contacts():
...
+
+    api_key = app.config['FRESHDESK_API_KEY']
+    url_request = requests.post("https://opensciencegrid.freshdesk.com/api/v2/tickets", 
+                                data = json.dumps(ticket),
+                                auth = (api_key, 'x'),
+                                headers = headers)
+
+    if url_request.status_code in [ 200, 201 ]:
+        json_return = url_request.json()
+        return_str = "Created Ticket {}".format(json_return['id'])
```
Comment (brianhlin) [inline src/app.py]: We should tell users that they'll receive an email shortly. Would it be possible to have this show up as a message on the form page? The page with the return_str is a little bare.

**Comment by matyasselmeci** [comment]
This seems like a useful tool but it's five years old and hasn't been merged. Sorry; closing.
